### ISSUE TITLE
`/sync` does not return "current state"

### ISSUE THREAD

**Issue by richvdh** [element-hq/synapse#16940]
Consider a DAG like this:

```
             E1
           ↗    ↖
          |      S2
          |
        --|---
          |
          E3
```

The client then initialsyncs, with `limit=1` (or maybe there are lots of events at E3), represented by the horizontal dashed line.

The "current state" (as defined by "what will be the state of the room if I send an event right now", which is what is important for encryption, etc) includes S2. However, the result from `/sync` does not include S2 (and will not, at least until there is an event in the DAG which joins the forks).

**Comment by richvdh** [comment]
In general, this seems like a major problem. As long as there is no event in the DAG which references `S2` as a prev_event, `S2` will not be served to the client. If `S2` is, say, a membership event, then this could lead to UTDs.

**Comment by kegsay** [comment]
Specficially this would in the common case cause a single UTD, as the act of sending the encrypted message would merge the forks and cause S2 to arrive on the client. The UTD cause would be not sharing room keys with the new user, for potentially a long time after they joined.

**Comment by richvdh** [comment]
> Specficially this would in the common case cause a single UTD,

Unfortunately I think it's considerably worse than that. Per https://github.com/element-hq/synapse/issues/16941, Synapse still won't send down S2 to the client even after the DAG is merged. I've proposed an improvement to that, but it still doesn't help clients which have lazy-loading enabled (which is basically all the ones we care about).

**Comment by richvdh** [comment]
This also extends to `/members` (at least when called with an `?at=` param, which is what you have to do for lazy-loading).

**Comment by richvdh** [comment]
> As long as there is no event in the DAG which references `S2` as a prev_event, `S2` will not be served to the client. If `S2` is, say, a membership event, then this could lead to UTDs.

It's not even clear that events that reference `S2` are sufficient to resolve this situation. Consider:

```
             E1
           ↗    ↖
          |      S2
          |       |
        --|-------|--
          |       |
          E3      |
          |       |
        --|-------|--
          |       |
          E4      |
          |       E5
          E6
```

What happens now? I'm pretty sure that S2 is not returned here.

Fundamentally, the expected behaviour is very poorly defined.

**Comment by erikjohnston** [comment]
The correct fix for all of this (probably) is to switch to using `current_state_delta_stream` table to track the state changes, rather than looking at the state at individual events. This is not a small piece of work.

**Comment by kegsay** [comment]
@erikjohnston would it help at all that at least for SSS we only care about returning the current state? We may not be able to fix it for sync v2 easily, but can we for SSS?

**Comment by erikjohnston** [comment]
> @erikjohnston would it help at all that at least for SSS we only care about returning the current state? We may not be able to fix it for sync v2 easily, but can we for SSS?

Yup! The SSS implementation actually tracks current state properly

**Comment by richvdh** [comment]
This will be fixed for `/sync` (but not `/members`, which arguably matters just as much from the POV of UTDs) by MSC4222.

**Comment by MadLittleMods** [comment]
As an update, in terms of `/sync`, [MSC4222](https://github.com/matrix-org/matrix-spec-proposals/pull/4222) has been accepted/merged and https://github.com/element-hq/synapse/pull/17888 is already in Synapse. Only missing piece is making it stable.

Is there a plan to make a similar MSC for `/members`?

**Comment by kegsay** [comment]
Surely you can get the right behaviour for `/members` today by just not specifying an `?at=` param? In which case it'll use the [current state](https://github.com/element-hq/synapse/blob/v1.137.0/synapse/handlers/message.py#L192-L195)?

**Comment by richvdh** [comment]
> Surely you can get the right behaviour for `/members` today by just not specifying an `?at=` param? In which case it'll use the [current state](https://github.com/element-hq/synapse/blob/v1.137.0/synapse/handlers/message.py?rgh-link-date=2025-09-05T07%3A05%3A10.000Z#L192-L195)?

The thing is, if you're a client trying to get the current membership list, you don't really want the *current state*, you want the state at a specific point in the sync stream, so that you can then maintain that list by tracking /sync responses. Hence in practice, clients use an `?at` param to populate their membership lists after a lazy-loading initialsync.
